// ng2-animate-on-scroll/__ivy_ngcc__/dist/index.js
"use strict";
var ɵngcc0 = require('@angular/core');
var ɵngcc1 = require('./src/animate-on-scroll.directive');
var ɵngcc2 = require('@angular/common');
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
function __export(m) {
    for (var p in m) if (!exports.hasOwnProperty(p)) exports[p] = m[p];
}
Object.defineProperty(exports, "__esModule", { value: true });
var core_1 = require("@angular/core");
var common_1 = require("@angular/common");
var src_1 = require("./src");
__export(require("./src/animate-on-scroll.directive"));
var AnimateOnScrollModule = /** @class */ (function () {
    function AnimateOnScrollModule() {
    }
    AnimateOnScrollModule_1 = AnimateOnScrollModule;
    AnimateOnScrollModule.forRoot = function () {
        return {
            ngModule: AnimateOnScrollModule_1,
            providers: [src_1.ScrollService]
        };
    };
    var AnimateOnScrollModule_1;
AnimateOnScrollModule.ɵmod = ɵngcc0.ɵɵdefineNgModule({ type: AnimateOnScrollModule });
AnimateOnScrollModule.ɵinj = ɵngcc0.ɵɵdefineInjector({ factory: function AnimateOnScrollModule_Factory(t) { return new (t || AnimateOnScrollModule)(); }, imports: [[
            common_1.CommonModule
        ]] });
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(AnimateOnScrollModule, { declarations: [ɵngcc1.AnimateOnScrollDirective], imports: [ɵngcc2.CommonModule], exports: [ɵngcc1.AnimateOnScrollDirective] }); })();
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(AnimateOnScrollModule, [{
        type: core_1.NgModule,
        args: [{
                imports: [
                    common_1.CommonModule
                ],
                declarations: [
                    src_1.AnimateOnScrollDirective
                ],
                exports: [
                    src_1.AnimateOnScrollDirective
                ]
            }]
    }], function () { return []; }, null); })();
    return AnimateOnScrollModule;
}());
exports.AnimateOnScrollModule = AnimateOnScrollModule;
