// ng2-scroll-to-el/__ivy_ngcc__/scrollTo.module.js
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { ScrollToDirective } from './scrollTo.directive';
import { ScrollToService } from './scrollTo.service';
import * as ɵngcc0 from '@angular/core';
var ScrollToModule = (function () {
    function ScrollToModule() {
    }
    ScrollToModule.forRoot = function () {
        return {
            ngModule: ScrollToModule
        };
    };
ScrollToModule.ɵmod = ɵngcc0.ɵɵdefineNgModule({ type: ScrollToModule });
ScrollToModule.ɵinj = ɵngcc0.ɵɵdefineInjector({ factory: function ScrollToModule_Factory(t) { return new (t || ScrollToModule)(); }, providers: [
        ScrollToService
    ], imports: [[
            CommonModule
        ]] });
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(ScrollToModule, { declarations: function () { return [ScrollToDirective]; }, imports: function () { return [CommonModule]; }, exports: function () { return [ScrollToDirective]; } }); })();
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(ScrollToModule, [{
        type: NgModule,
        args: [{
                declarations: [
                    ScrollToDirective
                ],
                imports: [
                    CommonModule
                ],
                exports: [
                    ScrollToDirective
                ],
                providers: [
                    ScrollToService
                ]
            }]
    }], function () { return []; }, null); })();
    return ScrollToModule;
}());
export { ScrollToModule };
/** @nocollapse */
ScrollToModule.ctorParameters = function () { return []; };
