// angular-typing-animation/__ivy_ngcc__/typing-animation.module.js
import { NgModule } from '@angular/core';
import { TypingAnimationDirective } from './typing-animation.directive';
import * as ɵngcc0 from '@angular/core';
var TypingAnimationModule = /** @class */ (function () {
    function TypingAnimationModule() {
    }
    /** @nocollapse */
    TypingAnimationModule.ctorParameters = function () { return []; };
TypingAnimationModule.ɵmod = ɵngcc0.ɵɵdefineNgModule({ type: TypingAnimationModule });
TypingAnimationModule.ɵinj = ɵngcc0.ɵɵdefineInjector({ factory: function TypingAnimationModule_Factory(t) { return new (t || TypingAnimationModule)(); } });
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(TypingAnimationModule, { declarations: function () { return [TypingAnimationDirective]; }, exports: function () { return [TypingAnimationDirective]; } }); })();
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(TypingAnimationModule, [{
        type: NgModule,
        args: [{
                declarations: [
                    TypingAnimationDirective
                ],
                exports: [
                    TypingAnimationDirective
                ]
            }]
    }], function () { return []; }, null); })();
    return TypingAnimationModule;
}());
export { TypingAnimationModule };
